### PR TITLE
Add network stats for Windows containers

### LINKED ISSUE #64901: Add network stats for Windows containers
PR #62266 adds kubelet stats for Windows pods/containers, e.g.

```json
{
  "node": {
    "nodeName": "12555k8s9000",
    "startTime": "2018-05-17T07:22:44Z",
    "cpu": {
      "time": "2018-05-17T07:25:49Z",
      "usageCoreNanoSeconds": 250800000000
    },
    "memory": {
      "time": "2018-05-17T07:25:49Z",
      "availableBytes": 5040005120,
      "usageBytes": 9516679168,
      "workingSetBytes": 3549458432,
      "rssBytes": 0,
      "pageFaults": 0,
      "majorPageFaults": 0
    },
    "fs": {
      "time": "2018-05-17T07:25:49Z",
      "availableBytes": 9437716480,
      "capacityBytes": 32212250624,
      "usedBytes": 22774534144
    },
    "runtime": {
      "imageFs": {
        "time": "2018-05-17T07:25:50Z",
        "availableBytes": 9437716480,
        "capacityBytes": 32212250624,
        "usedBytes": 22774534144,
        "inodesUsed": 0
      }
    }
  },
  "pods": [
    {
      "podRef": {
        "name": "iis-7b9b94fbd5-h7jzd",
        "namespace": "default",
        "uid": "2d442f1c-574d-11e8-b8d0-000d3aa00b4f"
      },
      "startTime": "2018-05-17T07:22:46Z",
      "containers": [
        {
          "name": "iis",
          "startTime": "2018-05-17T07:23:18Z",
          "cpu": {
            "time": "2018-05-17T07:25:51Z",
            "usageCoreNanoSeconds": 7390625000
          },
          "memory": {
            "time": "2018-05-17T07:25:51Z",
            "workingSetBytes": 43278336
          },
          "rootfs": {
            "time": "2018-05-17T07:25:51Z",
            "availableBytes": 9437716480,
            "capacityBytes": 32212250624,
            "usedBytes": 0
          },
          "logs": {
            "time": "2018-05-17T07:25:50Z",
            "availableBytes": 9437716480,
            "capacityBytes": 32212250624,
            "usedBytes": 22774534144,
            "inodesUsed": 0
          },
          "userDefinedMetrics": null
        }
      ],
      "cpu": {
        "time": "2018-05-17T07:23:18Z",
        "usageNanoCores": 0,
        "usageCoreNanoSeconds": 7390625000
      },
      "memory": {
        "time": "2018-05-17T07:25:51Z",
        "availableBytes": 0,
        "usageBytes": 0,
        "workingSetBytes": 43278336,
        "rssBytes": 0,
        "pageFaults": 0,
        "majorPageFaults": 0
      },
      "volume": [
        {
          "time": "2018-05-17T07:23:45Z",
          "availableBytes": 9462988800,
          "capacityBytes": 32212250624,
          "usedBytes": 22749261824,
          "inodesFree": 0,
          "inodes": 0,
          "inodesUsed": 0,
          "name": "default-token-r42mw"
        }
      ],
      "ephemeral-storage": {
        "time": "2018-05-17T07:25:51Z",
        "availableBytes": 9437716480,
        "capacityBytes": 32212250624,
        "usedBytes": 22774534144
      }
    }
  ]
}
```

But there's still no network stats for pods and containers. We should figure out how to fix that.

/sig windows
/kind bug

cc @michmike @PatrickLang @madhanrm @dineshgovindasamy

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

This PR adds network stats for Windows nodes and pods:

- To get node network stats, it invokes 'Get-NetAdapterStatistics | ConvertTo-Json' to fetch all adapters statistics
- To get pod stats, it adds `NetworkUsage` to CRI and gets data from docker API for Windows containers

Example stats after this PR:

```sh
$ curl -s localhost:8001/api/v1/nodes/<nodename>/proxy/stats/summary | jq '.node.network'
{
  "time": "2018-08-27T06:13:29Z",
  "name": "",
  "interfaces": [
    {
      "name": "vEthernet (nat) 2",
      "rxBytes": 29416,
      "rxErrors": 0,
      "txBytes": 4050473,
      "txErrors": 0
    },
    {
      "name": "vEthernet (Ethernet 2)",
      "rxBytes": 2613190,
      "rxErrors": 0,
      "txBytes": 4813970,
      "txErrors": 0
    },
    {
      "name": "Ethernet 2",
      "rxBytes": 3159653,
      "rxErrors": 0,
      "txBytes": 4972573,
      "txErrors": 0
    },
    {
      "name": "vEthernet (nat)",
      "rxBytes": 0,
      "rxErrors": 0,
      "txBytes": 0,
      "txErrors": 0
    }
  ]
}

$ curl -s localhost:8001/api/v1/nodes/<nodename>/proxy/stats/summary | jq '.pods[2].network'
{
  "time": "2018-08-27T06:13:54Z",
  "name": "b4897a3a-1501-4dd9-8bbb-2413619cf9d7",
  "rxBytes": 250962,
  "rxErrors": 0,
  "txBytes": 74626,
  "txErrors": 0,
  "interfaces": [
    {
      "name": "b4897a3a-1501-4dd9-8bbb-2413619cf9d7",
      "rxBytes": 250962,
      "rxErrors": 0,
      "txBytes": 74626,
      "txErrors": 0
    }
  ]
}
``` 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64901

**Special notes for your reviewer**:

The pod's adapter name is in format `b4897a3a-1501-4dd9-8bbb-2413619cf9d7`, which is not same within containers (e.g. `vEthernet`). It's ok now since only one adapter is supported now. 

The adapter name is got from docker API, we can fix that in the future docker releases if multiple adapters are required in the future.

**Release note**:

```release-note
Add network stats for Windows nodes and pods.
```

### CODE REVIEW

**Comment by dchen1107** [comment]
Here are some comments:

1) This PR changed CRI API by introducing container level network stats. We need more discussion on it.
2) More important, I am not sure if that is what we want since currently for Linux, Kubelet exports network stats at pod-level, not at container level.

**Comment by feiskyer** [comment]
> This PR changed CRI API by introducing container level network stats. We need more discussion on it.

Yep, this PR adds container network stats which is used for windows. This is because there's no cadvisor on Windows but there's already network stats from docker API.

> More important, I am not sure if that is what we want since currently for Linux, Kubelet exports network stats at pod-level, not at container level.

Windows sum up all container's stats to make pod level network stats. I was thinking two solutions for the windows stats originally, 

(1) add a new `ListSandboxStats()` interface in the CRI
(2) add network stats to current container stats and document it is only for windows

I decided (2) because it's much more intuitive to me.

cc @derekwaynecarr @Random-Liu @mrunalp @kubernetes/sig-node-pr-reviews for more comments on this.

**Comment by spiffxp** [comment]
/milestone v1.13
SIG Testing chair here, we're now in code freeze, so let's address next release

**Comment by feiskyer** [comment]
@derekwaynecarr @Random-Liu @mrunalp @kubernetes/sig-node-pr-reviews Any comments on this?

File: pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto [hunk 1]
```diff
@@ -1223,6 +1223,8 @@ message ContainerStats {
     MemoryUsage memory = 3;
     // Usage of the writeable layer.
     FilesystemUsage writable_layer = 4;
+    // Usage of the network interfaces.
```
Comment (yujuhong) [inline pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto]: I may have missed the previous discussions. Why do we need per-container (as opposed to per-pod) network stats?

**Review by yujuhong — CHANGES_REQUESTED** [review]

File: pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto (hunk 1 above)
Comment (feiskyer) [reply pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto]: See https://github.com/kubernetes/kubernetes/pull/67884#issuecomment-418580789. It is used to get network stats for Windows containers from runtime (e.g. docker).

**Comment by feiskyer** [comment]
Opened #70121 instead.
/hold

**Comment by dims** [comment]
per discussion in sig-arch, let's move this out to 1.14

/milestone v1.14

**Comment by feiskyer** [comment]
Close in favor of #70121.

**Pull request closed without merging**